File: Backend/VijayDairy.Postgres/UserRepository.cs
using System;
using System.Threading.Tasks;
using VijayDairy.Domain.Entities;
using VijayDairy.Persistence;

namespace VijayDairy.Postgres;

public class UserRepository : PostgresHelper, IUserRepository
{
    public UserRepository(string connectionString) : base(connectionString) { }

    private static User Map(System.Data.IDataReader reader) => new()
    {
        Id = reader.GetGuid(reader.GetOrdinal("id")),
        Email = reader.GetString(reader.GetOrdinal("email")),
        PasswordHash = reader.GetString(reader.GetOrdinal("password_hash")),
        Name = reader.GetString(reader.GetOrdinal("name")),
        Role = reader.GetString(reader.GetOrdinal("role")),
        CreatedAt = reader.GetDateTime(reader.GetOrdinal("created_at")),
    };

    public Task<User?> GetByEmailAsync(string email)
    {
        const string sql = "SELECT id, email, password_hash, name, role, created_at FROM users WHERE LOWER(email) = LOWER(@Email)";
        using var reader = ExecuteDataReader(sql, new[] { Param("@Email", email) });
        return Task.FromResult(reader.Read() ? Map(reader) : null);
    }

    public Task<User?> GetByIdAsync(Guid id)
    {
        const string sql = "SELECT id, email, password_hash, name, role, created_at FROM users WHERE id = @Id";
        using var reader = ExecuteDataReader(sql, new[] { Param("@Id", id) });
        return Task.FromResult(reader.Read() ? Map(reader) : null);
    }

    public Task<User> CreateAsync(User user)
    {
        const string sql = @"
            INSERT INTO users (email, password_hash, name, role)
            VALUES (@Email, @PasswordHash, @Name, @Role)
            RETURNING id, email, password_hash, name, role, created_at";
        using var reader = ExecuteDataReader(sql, new[]
        {
            Param("@Email", user.Email),
            Param("@PasswordHash", user.PasswordHash),
            Param("@Name", user.Name),
            Param("@Role", user.Role),
        });
        reader.Read();
        return Task.FromResult(Map(reader));
    }
}

File: Backend/src/repositories/userRepository.ts
import { getPool } from '../db';
import type { User } from '../types';

const COLUMNS = 'id, email, password_hash, name, role, created_at';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const map = (row: any): User => ({
  id: row.id,
  email: row.email,
  passwordHash: row.password_hash,
  name: row.name,
  role: row.role,
  createdAt: row.created_at,
});

export const userRepository = {
  async getByEmail(email: string): Promise<User | null> {
    const { rows } = await getPool().query(`SELECT ${COLUMNS} FROM users WHERE LOWER(email) = LOWER($1)`, [email]);
    return rows[0] ? map(rows[0]) : null;
  },

  async getById(id: string): Promise<User | null> {
    const { rows } = await getPool().query(`SELECT ${COLUMNS} FROM users WHERE id = $1`, [id]);
    return rows[0] ? map(rows[0]) : null;
  },

  async create(user: Pick<User, 'email' | 'passwordHash' | 'name' | 'role'>): Promise<User> {
    const { rows } = await getPool().query(
      `INSERT INTO users (email, password_hash, name, role) VALUES ($1, $2, $3, $4) RETURNING ${COLUMNS}`,
      [user.email, user.passwordHash, user.name, user.role],
    );
    return map(rows[0]);
  },
};
